### ISSUE TITLE
PolyMC lags when opening the instance page with long logs

### ISSUE THREAD

**Issue by flowln** [PolyMC/PolyMC#604]
### Operating System

Linux

### Version of PolyMC

PolyMC 1.2.2

### Description of bug

When the game has been running for a while, opening the instance page lags a lot (can take more than a minute hanging before opening for real). I use a HDD, so of course that's part of the problem (though people have said it also happens on SSDs). 

### Steps to reproduce

- Leave the game running for a while (may take more or less time depending on your storage device, mine lags after like, half hour in a modpack with 100-200 mods i think).
- Try to open the instance page
- Wait (possibly a lot) :P

### Suspected cause

When opening the instance page, the log is retrieved from disk in it's entirety, which can take a while with long logs / slow storage. Probably implementing a paging thing to retrieve only parts of logs when you're reading through them would help.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

**Comment by DioEgizio** [comment]
I can reproduce this, it's a bug that MultiMC has too. Temp workaround is using close the launcher after game window opens

**Comment by txtsd** [comment]
We can probably optimize how the log is read and displayed.

**Comment by Explodey-Wolf** [comment]
Explodey_Wolf: This especially happens when you have a log that is 23000 long. And if you need to see the log... Wait a few minutes!

**Comment by kthchew** [comment]
Is it possible to load the log on a separate thread? The log would still take a while to appear, but at least it wouldn’t freeze the launcher UI while it does.

**Comment by flowln** [comment]
> Is it possible to load the log on a separate thread? The log would still take a while to appear, but at least it wouldn’t freeze the launcher UI while it does.

It's definitely possible, but 1. How would the log appear as it's loading? and 2. PolyMC would still be using a lot of disk ops for nothing, possibly slowing down other things in the process.

**Comment by kthchew** [comment]
> > Is it possible to load the log on a separate thread? The log would still take a while to appear, but at least it wouldn’t freeze the launcher UI while it does.
> 
> It's definitely possible, but 1. How would the log appear as it's loading? and 2. PolyMC would still be using a lot of disk ops for nothing, possibly slowing down other things in the process.

For the first question, there would probably be some loading indicator (maybe just a spinning wheel or something) until it’s loaded. The second part might be true, but I think we just have to test that to know for sure.

Alternatively, does Qt have some kind of way to lazy load the log (as in, only load the part of the log that is visible)?

**Comment by flowln** [comment]
> For the first question, there would probably be some loading indicator (maybe just a spinning wheel or something) until it’s loaded. 

Doesn't change the fact that it'll take a lot of time to load in such instances though, and even if the part you're interested in is loaded already, it has to wait for the log in it's entirety to load before displaying it :<

> Alternatively, does Qt have some kind of way to lazy load the log (as in, only load the part of the log that is visible)?

We can seek to the end of the log file, and read only that last bit. If the user scrolls up, we seek to before, read the data and append it at the beginning. If we're careful with the boundaries and ranges, it should work out fine.

**Comment by ryanccn** [comment]
With this solution the scroll bar would be expanding constantly as we scroll up though, so maybe we can add some padding at the top? But I don't think we can easily get how many _lines_ a file has, so ¯\_(ツ)_/¯

I think the concept we're looking for here is **virtualized lists**. It's used quite a lot in data-heavy web applications, but applying it to Qt could be a pain. Maybe switch the logs to a `QListWidget` under the hood discreetly?

**Comment by flowln** [comment]
> With this solution the scroll bar would be expanding constantly as we scroll up though

It's how all other async-created lists behave in PolyMC though, so I don't think it'd be that big of a deal :slightly_smiling_face: 

> Maybe switch the logs to a QListWidget under the hood discreetly?

This doesn't seem very right... The log is not a list _per-se_, it's a text area. If we were to make it a list, it would be a list of what? of lines? Then there'd be too much items in the list and the overhead would be big. Of chunks of lines? Then what would be the size of those chunks? And what would be the behaviour of those in the boundaries compared to inside the items?

Idk, this seems to be over complicating the logs to me, but maybe it can work :p

**Comment by ryanccn** [comment]
yes `QListWidget` doesn't seem appropriate as there's no way to get how many lines in a file without reading the entire file
